File: LeetCode/Data-Str/Day-1/09-CQueue.cpp
//用两个栈实现队列，栈：先进后出；队列：先进先出
//stack_in：元素先进入stack_in栈中，判断stack_out是否为空，若为空，则删除时先将stack_in内的元素出栈并压入stack_out中，完成顺序颠倒，若stack_out不为空，即将栈顶元素pop
class CQueue {
    stack<int> stack_in, stack_out;
public:
    CQueue() {
        while(!stack_in.empty()){
            stack_in.pop();
        }
        while(!stack_out.empty()){
            stack_out.pop();
        }
    }
    
    void appendTail(int value) {
        stack_in.push(value);
    }
    
    int deleteHead() {
        if(stack_out.empty()){
            while(!stack_in.empty())
            {
                stack_out.push(stack_in.top());
                stack_in.pop();
            }
        }
        if(stack_out.empty()){
            return -1;
        }
        else {
            int deleteItem = stack_out.top();
            stack_out.pop();
            return deleteItem;
        }
    }
};

File: 剑指offer/链表/09_CQueue.py
class CQueue:

    def __init__(self):
        self.stack1 = []
        self.stack2 = []

    def appendTail(self, value: int) -> None:
        self.stack1.append(value)

    def deleteHead(self) -> int:
        # 如果stack2非空，则直接弹出stack2的头元素
        if self.stack2: return self.stack2.pop()

        # 如果stack1为空，stack2为空则整个队列为空，直接返回-1
        if not self.stack1: return -1

        # 当stack1非空时，先将元素放入stack2中再弹出头元素
        while self.stack1:
            val = self.stack1.pop()
            self.stack2.append(val)
        return self.stack2.pop()



# Your CQueue object will be instantiated and called as such:
# obj = CQueue()
# obj.appendTail(value)
# param_2 = obj.deleteHead()
